src/api/c/pinverse.cpp

```diff
@@ -0,0 +1,170 @@
+    Array<T> vT = createEmptyArray<T>(dim4(N, N));
+    svd<T, Tr>(sVec, u, vT, in);
+
+    // Cast s back to original data type for matmul later
+    // (since svd() makes s' type the base type of T)
+    Array<T> sVecCast = cast<T, Tr>(sVec);
+
+    Array<T> v = transpose(vT, true);
+
+    // Get reciprocal of sVec's non-zero values for s pinverse, except for
+    // very small non-zero values though (< relTol), in order to avoid very
+    // large reciprocals
+    double relTol = tol * static_cast<double>(max(M, N))
+                        * reduce_all<af_max_t, Tr, Tr>(sVec);
+    Array<T> relTolArr = createValueArray<T>(sVecCast.dims(), scalar<T>(relTol));
+    Array<T> ones = createValueArray<T>(sVecCast.dims(), scalar<T>(1.));
+    Array<T> sVecRecip = arithOp<T, af_div_t>(ones, sVecCast, sVecCast.dims());
+    Array<char> cond = logicOp<T, af_ge_t>(sVecCast, relTolArr,
+                                           sVecCast.dims());
+    Array<T> zeros = createValueArray<T>(sVecCast.dims(), scalar<T>(0.));
+    sVecRecip = createSelectNode<T>(cond, sVecRecip, zeros, sVecRecip.dims());
+    sVecRecip.eval();
+
+    // Make s vector into s pinverse array
+    Array<T> sPinv = diagCreate<T>(sVecRecip, 0);
+
+    Array<T> uT = transpose(u, true);
+
+    // Crop v and u* for final matmul later based on s+'s size, because
+    // sVec produced by svd() has minimal dim length (no extra zeroes).
+    // Thus s+ produced by diagCreate() will have minimal dims as well,
+    // and v could have an extra dim0 or u* could have an extra dim1
+    if (v.dims()[1] > sPinv.dims()[0]) {
+        vector<af_seq> seqs = {
+            {0., static_cast<double>(v.dims()[0] - 1), 1.},
+            {0., static_cast<double>(sPinv.dims()[0] - 1), 1.}
+        };
+        v = createSubArray<T>(v, seqs);
+    }
+    if (uT.dims()[0] > sPinv.dims()[1]) {
+        vector<af_seq> seqs = {
+            {0., static_cast<double>(sPinv.dims()[1] - 1), 1.},
+            {0., static_cast<double>(uT.dims()[1] - 1), 1.}
+        };
+        uT = createSubArray<T>(uT, seqs);
+    }
+
+    Array<T> out = matmul<T>(matmul<T>(v, sPinv, AF_MAT_NONE, AF_MAT_NONE),
+                             uT, AF_MAT_NONE, AF_MAT_NONE);
+
+    return out;
+}
+
+template<typename T>
+static inline af_array pinverse(const af_array in, const double tol)
+{
+    Array<T> inArray = getArray<T>(in);
+    uint batchSize = inArray.dims()[2];
+
+    if (inArray.ndims() < 3) {
+        return getHandle(pinverseSvd<T>(getArray<T>(in), tol));
+    }
+    else {
+        vector<Array<T> > outputs;
+        for (int i = 0; i < batchSize; ++i) {
+            vector<af_seq> seqs = {
+                {0., static_cast<double>(inArray.dims()[0] - 1), 1.},
+                {0., static_cast<double>(inArray.dims()[1] - 1), 1.},
+                {static_cast<double>(i), static_cast<double>(i), 1.}
+            };
+            Array<T> inSlice = createSubArray<T>(inArray, seqs);
```

Comment:
Can you make sure that this function is indeed making a subarray and not allocating and performing a copy? If it's copying an array then we should move this operation into the svd function and then iterate over the arrays there. You can just create a TODO if we are doing a copy and we can fix it at a later time. I suspect it is working as expected but I just want to make sure